### ISSUE TITLE
Local (multiprocessing) analysis run of bcbio overschedules main memory

### ISSUE THREAD

**Issue by schelhorn** [chapmanb/bcbio-nextgen#829]
I recently stumbled upon a condition which may be either a bug or intentional behaviour, and I was wondering if someone could clearify the issue for me. When running bcbio in local, multiprocessing mode (via `bcbio_nextgen.py sample.yaml -t local`), I noticed that bcbio consistently overschedules memory, resulting in the local node becoming unresponsive. 

More specifically, this occurred during `STAR` runs, which I had configured (via the `resource` section of the system yaml) to run on `cores: 4` and `memory: 8`, so using 32 GB of RAM at the maximum, per sample. My test node has 30GB of RAM, so with my configuration bcbio has to reduce the number of `STAR` cores to 3 to fit. This is indeed what happens: bcbio correctly deduces that the node has insufficient RAM for the maximum configuration and reduces the parallel job to 3 cores and 29GB of RAM (including `picard` piping and extra provisioning for the system, as deduced by [bcbio.distributed.resource](https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/distributed/resources.py)). So far, so good.

However, I have all in all 4 such jobs (since I have 4 samples as inputs in my sample configuration yaml), and what happens now is somewhat peculiar: bcbio runs all 4 of them in parallel on the local node via multiprocessing. Obviously, the node quickly dies trying to allocating 4 times its physical memory and swapping out like crazy. 

Therefore, I wonder, does local multiprocessing support resource _scheduling_ at all? Given that bcbio already has all information to avoid over-allocating the node, is it would seem intuitive that jobs are scheduled to run one after the other (of in parallel if their combined resource allocation fits within the node's resources). Or is local processing just considered to be a test bed for running an analysis of single samples, and users a expected to use a proper cluster manager such as SGE for parallel scheduling? In that case this should probably be mentioned in the docs or people may assume that they can use bcbio without a cluster. Or, last, did I err on my side and just badly configured local processing?

**Comment by roryk** [comment]
Hi Eric,

We support doing the processing just on the local machine, so it should be doing the right thing and not trying to run everything at once. Could you stick the log files up somewhere so we could look and try to figure out what is going on?

**Comment by schelhorn** [comment]
Hi Rory,

thanks for the offer, However, I fear there is not much in the logs that is informative. But I dug deeper and believe there may be a bug that may generally break multicore with large-mem jobs under certain conditions (or else, I have misunderstood how the number of jobs are distributed by bcbio, which may well be the case). 

It seems to me as if in the multicore setting the job demand is first computed on available cores alone and memory scaling is only incorporated later, about [here in bcbio.distributed.resources](https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/distributed/resources.py#L161). So for a job like STAR, with my adapted requirements of `cores=4` and `mem=8` per job (since a human genome for STAR requires about 30GB of RAM,  `4 * 8 > 30`), and a hypothetical 8 core 60GB machine (what I am now using for testing, more RAM than in my last post above), bcbio will first find the number of jobs per item (an item being a set of jobs run in parallel) by looking on the cores alone. So if will set this to two jobs (`2 * 4 <= 8`) and then look at the RAM to fit into system memory. 

That's where things may go wrong (or so I believe, based on my rather shallow review), since the code linked above will not be reached (because `cores_per_job` is `4`, not `1` for STAR), and so the _number of jobs_ are never adapted. Instead, they stay at `2` and thus overschedule the memory (two STAR indexed weigh more than the system RAM). 

Adapting the number of cores per process in the subsequent lines does not help: `_scale_cores_to_memory` thinks it's fine to have `4` cores for STAR since the system has `8` cores, and it does not look at the number of jobs that were scheduled (which are `2`). Even if the system would have only `3` cores, this would only lead to a insufficient reduction by `_scale_cores_to_memory` (from `4` to `3`) since it does not optimize towards the number of jobs that are scheduled, so the system is still overscheduled, isn't it? (Even then, STAR requires a constant-size index of 30GB RAM, since that does not scale with the number of cores anyway, but that is a separate issue). 

Well, then again, I may be terribly wrong :)

**Comment by lpantano** [comment]
Hi @schelhorn,

amazing detecting work. I was wondering what happens if you add to the command line `-n 4`? still scheduling two jobs. Maybe this will help to narrow the error a little bit more. But it seems the bug it is around those functions you mentioned.

**Comment by chapmanb** [comment]
Sven-Eric and Lorena;
Thanks so much for the awesome detective work. Apologies for the issue and for having to dig through that code. You're right on that this was no considering the local case where you are multi-core and have multiple jobs at the same time. I added a catch for this case where it will scale down jobs based on memory. This does the right thing for your case and will hopefully avoid the overscheduling issue you're seeing.

Memory dominated jobs like STAR are definitely hard to work with in general, since we don't have a great way to specify and scale these right now. Longer term it would be worth having better handling of this but practically hopefully this keeps you going.

**Comment by schelhorn** [comment]
Thanks a bunch for the quick fix, Brad. Concerning the constant memory requirements of STAR I believe an additional memory option or flag in the resource YAMLs would be best. There even is a similar option hardcoded for the aligners to ensure minimum memory, but also per core. But that has time.

**Comment by lpantano** [comment]
@schelhorn,
could you run with the last version to see if is still working for you? I did another commit after @chapmanb because I was getting only 1 job for steps using tools with multiple cores and a lot of memory. The last commit gets that working again, and it should still work for you.

please, if not, let us known and we will try to fix it again. 

thanks

**Comment by chapmanb** [comment]
Brilliant catch, thanks Lorena. I was thinking about this more as well and realized I should have been restricting it only to local jobs, so added that in as well. Please let me know if you run into any other issues at all with this.
